src/drivers/imx/esai.c

```diff
@@ -3,38 +3,413 @@
+	 * TODO use ESAI params instead of hardcode to compute the clock divider
+	 * settings.
+	 */
+	dai_update_bits(dai, REG_ESAI_TCCR, ESAI_xCCR_xFP_MASK,
+			ESAI_xCCR_xFP(1));
+	dai_update_bits(dai, REG_ESAI_TCCR, ESAI_xCCR_xPSR_MASK,
+			ESAI_xCCR_xPSR_DIV8);
+
+	/* Remove ESAI personal reset */
+	dai_update_bits(dai, REG_ESAI_xCR(0), ESAI_xCR_xPR, 0);
+	dai_update_bits(dai, REG_ESAI_xCR(1), ESAI_xCR_xPR, 0);
+
+	/* Configure ESAI pins, enable them all */
+	dai_update_bits(dai, REG_ESAI_PRRC, ESAI_PRRC_PDC_MASK,
+			ESAI_PRRC_PDC(ESAI_GPIO));
+	dai_update_bits(dai, REG_ESAI_PCRC, ESAI_PCRC_PC_MASK,
+			ESAI_PCRC_PC(ESAI_GPIO));
 	return 0;
 }
 
+static void esai_start(struct dai *dai, int direction)
+{
+	/* FIFO enable */
+	dai_update_bits(dai, REG_ESAI_xFCR(direction), ESAI_xFCR_xFEN_MASK,
+			ESAI_xFCR_xFEN);
+
+	/* To prevent channel swap issues, write a sample per channel. The ESAI
+	 * can begin transmitting these samples while it is requesting for a
+	 * DMA transfer, before said DMA transfer actually puts more samples
+	 * into the ESAI FIFO. Only needed on playback/transmit.
+	 */
+	if (!direction) {
+		/* TODO adjust this when we have more than 2 channels;
+		 * we need to write a word per active channel.
+		 */
+		dai_write(dai, REG_ESAI_ETDR, 0);
+		dai_write(dai, REG_ESAI_ETDR, 0);
+	}
+
+	/* Enable one transmitter or receiver */
+	dai_update_bits(dai, REG_ESAI_xCR(direction),
+			ESAI_xCR_xE_MASK(direction), ESAI_xCR_xE(direction, 1));
+
+	/* Configure time slot registers, enable two time slots for two
+	 * channels.
+	 *
+	 * This actually begins playback/record.
+	 *
+	 * The order of setting xSMB first and xSMA second is required to
+	 * correctly start playback/record; setting them in reverse order may
+	 * cause significant channel swap issues when using more than 16
+	 * channels.
+	 */
+	dai_update_bits(dai, REG_ESAI_xSMB(direction), ESAI_xSMB_xS_MASK,
+			ESAI_xSMB_xS_CHANS(2));
+
+	dai_update_bits(dai, REG_ESAI_xSMA(direction), ESAI_xSMA_xS_MASK,
+			ESAI_xSMA_xS_CHANS(2));
+}
+
+static void esai_stop(struct dai *dai, int direction)
+{
+	/* Disable transmitters/receivers */
+	dai_update_bits(dai, REG_ESAI_xCR(direction),
+			direction ? ESAI_xCR_TE_MASK : ESAI_xCR_RE_MASK, 0);
+
+	/* Turn off pins */
+	dai_update_bits(dai, REG_ESAI_xSMA(direction), ESAI_xSMA_xS_MASK, 0);
+	dai_update_bits(dai, REG_ESAI_xSMB(direction), ESAI_xSMB_xS_MASK, 0);
+
+	/* disable and reset FIFO */
+	dai_update_bits(dai, REG_ESAI_xFCR(direction),
+			ESAI_xFCR_xFR | ESAI_xFCR_xFEN, ESAI_xFCR_xFR);
+	dai_update_bits(dai, REG_ESAI_xFCR(direction), ESAI_xFCR_xFR, 0);
+}
+
 static int esai_trigger(struct dai *dai, int cmd, int direction)
 {
+	tracev_esai("ESAI: trigger");
+
+	switch (cmd) {
+	case COMP_TRIGGER_START:
+		esai_start(dai, direction);
+		break;
+	case COMP_TRIGGER_STOP:
+		esai_stop(dai, direction);
+		break;
+	case COMP_TRIGGER_PAUSE:
+		esai_stop(dai, direction);
+		break;
+	case COMP_TRIGGER_RELEASE:
+		//esai_start(dai, direction);
```

Comment:
To be removed?

Reply:
I'm not sure how the Release trigger is supposed to work. Guess I'll remove it because it works but I want to understand why it works without anything here.

Reply:
It's weird, because you are stopping it on PAUSE. If it's working after RELEASE, then either you don't actually stop it or something else enables it.

Reply:
Well on release somehow I'm getting the RELEASE and START triggers, and on pause I'm getting the STOP trigger. Somehow the flow is very weird with both the DMA and DAI drivers.

Reply:
(also no, pause/release doesn't _actually_ work, on the kernel I'm testing with the "pause" function says the hardware doesn't support pause). So I guess I'll keep it for now instead of removing.

Reply:
Yes, I forgot about it. For release you can additionally restore your previous position or do some other stuff if you need to. Apparently you doesn't.

Reply:
Position restore: I don't think I'm doing it with EDMA either, in fact I think I'm even resetting the FIFO so the pause/release flow is actually incorrect. But I currently cannot test it, regular playback works and I should come back to this when we want pause/release to work.

Should I just add a "todo" here for pause/release so that future me is aware there is an issue?